url29ecd5d728_geodb_db3dcore_tgz-pJ8-de_uos_igf_db3d_dbms_model3d_SegmentNetBuilderJ8/src/de/uos/igf/db3d/dbms/geom/GeomUtils.java

```diff
@@ -0,0 +1,1416 @@
+	 *            a non-NaN- double value which may also be infinite
+	 * @param d2
+	 *            another non-NaN- double value which may also be infinite
+	 * @return the numeric order of the specified arguments in the same fashion
+	 *         as a comparator does.
+	 * @throws IllegalArgumentException
+	 *             - if the given doubles are NaN values and cannot be compared.
+	 */
+	public static int doubleCompare(double d1, double d2) {
+
+		if (d1 > d2) {
+			return +1;
+		}
+		if (d1 < d2) {
+			return -1;
+		}
+		if (d1 == d2) {
+			return 0;
+		}
+
+		throw new IllegalArgumentException(Db3dSimpleResourceBundle
+				.getString("db3d.geom.comparenan"));
+
+	}
+
+	/**
+	 * Computes the (square of) the distance between the affine hulls of the two
+	 * given point sets and eventually also the affine coefficients of points at
+	 * minimal distance.
+	 * 
+	 * If a coeffs parameter is specified the corresponding coefficients of the
+	 * affine combination of a closest point is specified. {@link http
+	 * ://en.wikipedia.org/wiki/Affine_combination}
+	 * 
+	 * The algorithm solves the optimization problem minimize length(coeffsA[0]
+	 * * a[0] + ... + coeffsA[n] * a[n] - coeffsB[0]*b[0] - ... -
+	 * coeffsB[m]*b[m]) as a function in the coefficients respecting the
+	 * conditions coeffsA[0] + ... + coeffsA[n] = 1 coeffsB[0] + ... +
+	 * coeffsB[m] = 1.
+	 * 
+	 * The linear equation system kktM * kktX = kktY is iteratively constructed
+	 * using the Lagrangians of the previous solution and at each iteration step
+	 * adding the point which is most "inclined" towards the other currently
+	 * considered subset. {@link http
+	 * ://en.wikipedia.org/wiki/Quadratic_programming}
+	 * 
+	 * 
+	 * @param a
+	 *            an array of points defining an affine space clA
+	 * @param b
+	 *            an array of points defining another affine space clB
+	 * @param coeffsA
+	 *            if not null, the affine coefficients for a point pa at minimal
+	 *            distance to clB, i.o.w.
+	 *            <code>pa = sum coeffsA[i]*a[i] for i in 0..a.length</code>,
+	 *            will be stored here. If the specified array is too short a
+	 *            ArrayIndexOutOfBoundsException will occur.
+	 * @param coeffsB
+	 *            if not null, the affine coefficients for a point pb at minimal
+	 *            distance to clA, i.o.w.
+	 *            <code>pb = sum coeffsB[j]*b[j] for j in 0..b.length</code>,
+	 *            will be stored here. If the specified array is too short a
+	 *            ArrayIndexOutOfBoundsException will occur.
+	 * @param scOp
+	 *            the Scalar operator which compares double values up to a
+	 *            specified error.
+	 * @return the square of the distance between the affine closure of a and
+	 *         the affine closure of b. TODO: Upload the file containing the
+	 *         description of the algorithmic technique used here.
+	 */
+	static double affineDistance(double[][] a, double[][] b, double[] coeffsA,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 38. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.